packages/core/src/utils.ts

```diff
@@ -259,6 +263,137 @@ export async function* maskStreamTags(
+    (!('type' in tool) || tool.type === undefined || tool.type === 'function')
+  );
+}
+
+interface ToolOptions {
+  name: string;
+  runId?: string;
+  threadId?: string;
+  resourceId?: string;
+  logger: Logger;
+  description?: string;
+  mastra?: MastraPrimitives;
+  memory?: MastraMemory;
+  agentName?: string;
+}
+
+type ToolToConvert = VercelTool | ToolAction<any, any, any, any>;
+
+interface LogOptions {
+  agentName?: string;
+  toolName: string;
+  tool?: ToolToConvert;
+  type?: 'tool' | 'toolset';
+}
+
+interface LogMessageOptions {
+  start: string;
+  error: string;
+}
+
+function createLogMessageOptions({ agentName, toolName, tool, type }: LogOptions): LogMessageOptions {
+  // If no agent name, use default format
+  if (!agentName) {
+    return {
+      start: `Executing tool ${toolName}`,
+      error: `Failed tool execution`,
+    };
+  }
+
+  const prefix = `[Agent:${agentName}]`;
+  const vercelPrefix = isVercelTool(tool) ? 'Vercel ' : '';
+  const toolType = type === 'toolset' ? 'toolset' : 'tool';
+
+  return {
+    start: `${prefix} - Executing ${vercelPrefix}${toolType} ${toolName}`,
+    error: `${prefix} - Failed ${vercelPrefix}${toolType} execution`,
+  };
+}
+
+function createExecute(tool: ToolToConvert, options: ToolOptions, logType?: 'tool' | 'toolset') {
+  const { logger, ...rest } = options;
+
+  const { start, error } = createLogMessageOptions({
+    agentName: options.agentName,
+    toolName: options.name,
+    tool,
+    type: logType,
+  });
+
+  const execFunction = async (args: any, execOptions: any) => {
+    if (isVercelTool(tool)) {
+      return tool?.execute?.(args, execOptions) ?? undefined;
+    }
+    return (
+      tool?.execute?.(
+        {
+          context: args,
+          threadId: options.threadId,
+          resourceId: options.resourceId,
+          mastra: options.mastra,
+          memory: options.memory,
+          runId: options.runId,
+        },
+        execOptions,
+      ) ?? undefined
+    );
+  };
+
+  return async (args: any, execOptions?: any) => {
+    try {
+      logger.debug(start, { ...rest, args });
+      return await execFunction(args, execOptions);
+    } catch (err) {
+      logger.error(error, { ...rest, error: err, args });
+      throw err;
+    }
+  };
+}
+
+/**
+ * Converts a Vercel Tool or Mastra Tool into a CoreTool format
+ * @param tool - The tool to convert (either VercelTool or ToolAction)
+ * @param options - Tool options including Mastra-specific settings
+ * @param logType - Type of tool to log (tool or toolset)
+ * @returns A CoreTool that can be used by the system
+ */
+export function makeCoreTool(tool: ToolToConvert, options: ToolOptions, logType?: 'tool' | 'toolset'): CoreTool {
+  // Helper to get parameters based on tool type
+  const getParameters = () => {
+    if (isVercelTool(tool)) {
+      return tool.parameters instanceof z.ZodType
+        ? tool.parameters
+        : resolveSerializedZodOutput(jsonSchemaToZod(tool.parameters));
```

Comment:
A comment on how this works would be good. I'm not sure in what case a vercel tool will have a string schema. Is this more of an internal thing?

Reply:
Vercel's tool format either accepts a zod schema or a jsonschema. So this check checks if parameters are zod (which it will leave alone) or json schema, which it will then convert.

Reply:
Added a comment